browser-state: Move state handling to generic lib-state component.

pocmo opened the pull request:
### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

NotWoods commented on components/lib/state/src/main/java/mozilla/components/lib/state/Store.kt:
```diff
@@ -0,0 +1,108 @@
+    private val reducer: Reducer<S, A>
+) {
+    private val storeContext = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+    private val storeScope = CoroutineScope(storeContext)
```
Should `Store` use `CoroutineScope` as a delegate instead?

NotWoods commented on components/lib/state/src/main/java/mozilla/components/lib/state/Store.kt:
```diff
@@ -0,0 +1,108 @@
+) {
+    private val storeContext = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+    private val storeScope = CoroutineScope(storeContext)
+    private val subscriptions: MutableList<Subscription<S, A>> = mutableListOf()
```
Use a `MutableSet` instead since the subscriptions are removed by value, not index

NotWoods commented on components/lib/state/src/main/java/mozilla/components/lib/state/ext/StoreExtensions.kt:
```diff
@@ -60,10 +62,10 @@ private class LifecycleBoundObserver(
  * View.OnAttachStateChangeListener implementation to bind an observer to a View.
  */
-private class ViewBoundObserver(
+private class ViewBoundObserver<S : State, A: Action>(
```
Since `LifecycleBoundObserver` and `ViewBoundObserver` are both bindings, perhaps they should be called `Binding` instead?

NotWoods commented on components/lib/state/src/main/java/mozilla/components/lib/state/ext/StoreExtensions.kt:
```diff
@@ -45,10 +47,10 @@ fun BrowserStore.observe(
+) : LifecycleObserver, Store.Subscription.Binding {
     @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
     override fun unbind() {
         subscription.unsubscribe()
```
This looks like infinite recursion

NotWoods commented on components/lib/state/src/main/java/mozilla/components/lib/state/helpers/Helpers.kt:
```diff
@@ -0,0 +1,41 @@
+@Suppress("FunctionNaming")
+fun <S : State, T> onlyIfChanged(
+    onMainThread: Boolean = false,
+    map: (S) -> T?,
```
It think it makes more sense to extract mapping to its own helper.

grigoryk commented on components/lib/state/src/main/java/mozilla/components/lib/state/Store.kt:
```diff
@@ -0,0 +1,108 @@
+    initialState: S,
+    private val reducer: Reducer<S, A>
+) {
+    private val storeContext = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
```
not to bike-shed, but this is a dispatcher, not context, right?

grigoryk commented on components/lib/state/src/main/java/mozilla/components/lib/state/Store.kt:
```diff
@@ -0,0 +1,108 @@
+        var binding: Binding? = null
+
+        fun unsubscribe() {
+            store.removeSubscription(this)
```
This may be a no-op, if the store doesn't have this `subscription` anymore, right? Currently the only way to remove a subscription from the list is via this method, but perhaps that should either be documented, or somehow enforced?

grigoryk commented on components/lib/state/src/main/java/mozilla/components/lib/state/Store.kt:
```diff
@@ -0,0 +1,108 @@
+            observer.invoke(currentState)
+        }
+
+        return subscription
```
It's possible for the `subscription` to outlive the `store`, right? Lifetimes of these two aren't currently coordinated at all, once you return the `subscription`. And subscription has a reference to the `store`, which is opaque to users of the subscription. I don't think this sort of a class would actually compile in Rust, where this stuff is enforced.

NotWoods replied:
This is similar to how redux handles subscriptions, but using a `WeakRef` for the store might be better.

NotWoods commented:
Can supplying a binding be required? I think defaulting to using bindings is better and a "ForeverBoundObservable" class could be used for cases where it's needed.

pocmo commented:
Updated the PR with multiple small changes and the following bigger changes:
* I removed the `receiveInitialState` flag. It looks like we will always want to get the current state passed to the observer. If there will be a reason to not do that then we can add it later on. But for now removing this makes the code simpler.
* I renamed the default `observe` method to `observeManually` and added a `@CheckResult` annotation, which will cause a warning if the returned `Subscription` is not used by calling code. This hopefully makes it more obvious that you are on your own if you use that method and makes it less likely to use it accidentally. The extension methods are still called `observe` (with an `observeForever` added) and do not return a `Subscription` anymore since that is managed for you.
* The `Store` reference in `Subscription` is now a weak reference only.

pocmo commented:
Landing to unblock the follow-up work.
